code/espurna/ntp.ino

```diff
@@ -2,264 +2,343 @@
+void _ntpBrokerSchedule(int offset) {
+    _ntp_broker_timer.once_scheduled(offset ?: 1, _ntpBrokerCallback);
 }
 
-#endif
+void _ntpBrokerCallback() {
 
-void _ntpLoop() {
+    if (!ntpSynced()) {
+        _ntpBrokerSchedule(60);
+        return;
+    }
 
-    // Disable ntp sync when softAP is active. This will not crash, but instead spam debug-log with pointless sync failures.
-    if (!wifiConnected()) return;
+    const auto ts = now();
 
-    if (_ntp_configure) _ntpConfigure();
+    // current  time and formatter string is in local TZ
+    tm local_tm;
+    localtime_r(&ts, &local_tm);
 
-    // NTPClientLib will trigger callback with sync status
-    // see: NTPw.onNTPSyncEvent([](NTPSyncEvent_t error){ ... }) below
-    if (_ntp_want_sync) {
-        _ntp_want_sync = false;
-        NTPw.getTime();
-    }
+    int now_hour = local_tm.tm_hour;
+    int now_minute = local_tm.tm_min;
 
-    // Print current time whenever configuration changes or after successful sync
-    if (_ntp_report) _ntpReport();
+    static int last_hour = -1;
+    static int last_minute = -1;
 
-    #if BROKER_SUPPORT
-        _ntpBroker();
-    #endif
+    String datetime;
+    if ((last_minute != now_minute) || (last_hour != now_hour)) {
+        datetime = ntpDateTime(&local_tm);
+    }
 
-}
+    // notify subscribers about each tick interval (note that both can happen simultaneously)
+    if (last_hour != now_hour) {
+        last_hour = now_hour;
+        NtpBroker::Publish(NtpTick::EveryHour, ts, datetime.c_str());
+    }
 
-// TODO: remove me!
-void _ntpBackwards() {
-    moveSetting("ntpServer1", "ntpServer");
-    delSetting("ntpServer2");
-    delSetting("ntpServer3");
-    int offset = getSetting("ntpOffset", NTP_TIME_OFFSET);
-    if (-30 < offset && offset < 30) {
-        offset *= 60;
-        setSetting("ntpOffset", offset);
+    if (last_minute != now_minute) {
+        last_minute = now_minute;
+        NtpBroker::Publish(NtpTick::EveryMinute, ts, datetime.c_str());
     }
-}
 
-// -----------------------------------------------------------------------------
+    // try to autocorrect each invocation
+    _ntpBrokerSchedule(60 - local_tm.tm_sec);
 
-bool ntpSynced() {
-    #if NTP_WAIT_FOR_SYNC
-        // Has synced at least once
-        return (NTPw.getFirstSync() > 0);
-    #else
-        // TODO: runtime setting?
-        return true;
-    #endif
 }
 
-String ntpDateTime(time_t t) {
-    char buffer[20];
-    snprintf_P(buffer, sizeof(buffer),
-        PSTR("%04d-%02d-%02d %02d:%02d:%02d"),
-        year(t), month(t), day(t), hour(t), minute(t), second(t)
-    );
-    return String(buffer);
-}
+#endif
 
-String ntpDateTime() {
-    if (ntpSynced()) return ntpDateTime(now());
-    return String();
+void _ntpSetTimeOfDayCallback() {
+    _ntp_synced = true;
+    _ntp_last = time(nullptr);
+    #if BROKER_SUPPORT
+    static bool once = true;
+    if (once) {
+        _ntpBrokerSchedule(60 - second(_ntp_last));
```

Comment:
Using the same scheduling policy here as in _ntpBrokerCallback(). Mostly to avoid running something on 58th second and 00th after that, always force the timer to expire at 0 seconds.